### ISSUE TITLE
Allow users to choose if scripts should be forced to be statically typed (or not)

### ISSUE THREAD

**Issue by dreamsComeTrue** [godotengine/godot-proposals#173]
**Describe the project you are working on:**

Godot source code itself / RPG-type game in 2D

**Describe the problem or limitation you are having in your project:**

Having lot of GDScript files, I am not forced anyhow to maintain consistency across them.
Thus - I sometimes use optional typing, sometimes not.
I would love to have some sort of switch (on Editor Settings for example) that will force me
to remember to put types (or not) in Script files.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**

Having scripts checked for using optional typing across whole files would set harmony across whole project.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**

![image](https://user-images.githubusercontent.com/1110337/67208158-49228380-f415-11e9-9e20-26f6b7599b63.png)

**Describe implementation detail for your proposal (in code), if possible:**

Probably (apart from Editor Settings changes) that would need a lot of modifications of script_text_editor.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**

Only, if team/developers agree to have one appointed way of writing scripts

**Is there a reason why this should be core and not an add-on in the asset library?:**

It should have access to Script Editor itself (or some kind of background checker/code linter) - and as long as script_text_editor is a plugin itself - it could be also considered as a addon.

**Comment by KoBeWi** [comment]
It's basically like warning about unsafe lines, no?

**Comment by ludorverr** [comment]
When the planned speedboost for using typed-variables is implemented, I would definitely want to have this feature so everything can be consistently optimized.

**Comment by aaronfranke** [comment]
IIRC, one big issue with this currently is that there are some pieces of code you can't write with static typing, such as if you need nullable value types.

**Comment by Calinou** [comment]
If we want this to be implemented, we need to have an `any` type hint first, so that you can explicitly mark values as dynamaically typed.

**Comment by bojidar-bg** [comment]
Note that such a setting should likely live in the Project Settings, since it should be the same for all team members working on a project.

**Comment by Jummit** [comment]
This looks like it could be a by default disabled warning.

**Comment by dreamsComeTrue** [comment]
> It's basically like warning about unsafe lines, no?

Am I missing something - is there a magic switch for that? What I would like to achieve is one big list of all places in scripts where there is no type information which is dynamically updated as I am editing them. Currently, there is an Errors tab on Debugger pane, and it populates kinda similar data, but that happens only after I run the game (more like a debug output of all code places which were executed and had some problems). It is not being updated 'on the fly'.

**Comment by KoBeWi** [comment]
> Am I missing something - is there a magic switch for that?

There's none. "Safe" lines are code with unambiguous and strictly static typing (so no dynamic stuff). Their numbers are marked in green-ish color. I guess if we implemented a warning about unsafe lines, it could be enough for this proposal. Although there are some cases where some lines are marked as "unsafe" when they shouldn't.

**Comment by fab918** [comment]
I don't know if I should post a dedicated proposal but I want to extend your proposal to resolve your issue: keep the best of the two-word and switch to a (semi) typed language.

There are practically only advantages:

- auto-complete
- keep consistency with tutorials and examples
- avoid bugs
- increase the maintainability (especially for a serious project)
- performance (one of the main complain about GDScript, people don't want to learn GDNative)
- etc.

The main drawback is the extra syntax, but that's why I say "semi typed", the magic of GDscript is to be able to make type inferences in most of the situation. In other words, **the language can be semi-typed without writing any types in code in most of the case**, so why keep this inconsistency?

**Comment by aaronfranke** [comment]
> @fab918: The main drawback is the extra syntax

The main drawback is breaking backwards compatibility. This likely won't happen even though I personally think it's a good idea.

**Comment by fab918** [comment]
@aaronfranke Godot 4 will be a major release, major release can have breaking changes if the benefits worth it. Maybe a survey on this, explaining the stakes, can be a good option!?

Moreover, I don't know if it's a good idea but if you set an `any` type like Typescript, you can probably convert old code to new one without breaking change.

It's a shame to have to choose between fun and fast syntax (GDscript) or robust language (c++/c#) when GDscript can be both.

**Comment by Calinou** [comment]
> Godot 4 will be a major release, major release can have breaking changes if the benefits worth it. Maybe a survey on this, explaining the stakes, can be a good option!?

We already have some compatibility breakages planned, we don't want to add too much of them to not end up in a situation like Godot 3 where many people never migrated to it.

**Comment by fab918** [comment]
> I also proposed adding an `any` keyword above. Combining it with a project setting that requires everything to have a type _IF THE TYPE IS NOT INFERABLE_ annotation is likely sufficient (like TypeScript's noImplicitAny).

I added few words to your quote I think they are important. If your assertion on performance optimization is correct, it's a good balanced solution (I think it's important if this box is checked by the default to promote best practice for users, especially beginners).

**Comment by Calinou** [comment]
@fab918 Of course, the point of the setting isn't to forbid type inference. I like my `:=` too :slightly_smiling_face:

**Comment by fab918** [comment]
> That said, I'd say it's a bad idea to make = automatically infer the type as it'd change how the language would behave depending on project settings.

we are near to be 100% agree :)

I understand the principle, but forasmuch it's only for a **hint**, is it a really problem? Because the alternative is to be forced to use an heavy/ugly/usless syntax 
`:=`, especially if this setting become the standard in the community, what I'm pretty sure it will be when the performance optimization will be available). And in term of principles, a useless syntax is against the spirit of GdScript / Python.)

**Comment by Calinou** [comment]
> Because the alternative is to be forced to use an heavy/ugly/useless syntax

This syntax is established in many languages such as Go. I don't think it's so alien to people, especially as it's well-docoumented in the [static typing tutorial](https://docs.godotengine.org/en/latest/getting_started/scripting/gdscript/static_typing.html#how-to-use-static-typing).

**Comment by fab918** [comment]
> This syntax is established in many languages such as Go. I don't think it's so alien to people, especially as it's well-docoumented in the static typing tutorial.

It makes sense in a dynamical context to do `:=`but this setting said that the context is a typed one,  and in a typed one, we want always do `:=`, so this operator becomes useless/redondant

> There's "Explicit is better than implicit." in the Zen of Python still 😉

You missed the sentence before :D "Beautiful is better than ugly." and the next one "Simple is better than complex."

**Comment by Calinou** [comment]
@fab918 My point is that we should avoid making language behavior depend on project settings. There are many practical issues with this:

- Third-party assets may stop working if you import them into your own project with different settings.
- This breaks existing tutorials and demos that people can find online.
- It makes it harder to find solutions to scripting problems, as different people may effectively be using a different version of GDScript without realizing it.

**Comment by fab918** [comment]
The setting will break all non-typed example/tutorial in any case. That's why I suggested making this setting a default, to promote this direction for the futur.

Maybe an option to erase/add types in the editor can help to switch easily.

Anyway, I can live with that and I don't want to extend too much the discussion for a detail, **I will be more than happy with the `any` keyword and this setting**

**Comment by Wavesonics** [comment]
I think doing it on a per file level makes more sense. Adding a keyword like `strict` to the top of a file enforces static typing for that file. That way you can mix and match. A non strict project can include strict code from the asset store, and receve benefit from it being strict, or visa versa.

This is some what similar to what JavaScript does: https://www.w3schools.com/js/js_strict.asp

The other advantage of per file, is this obviously impacts how the code is written and read, so putting it in the code to be read is nice. Having it as a project setting decouples it from the code, you have to go check some where else to have all of the information. And if you include code from the asset store you can't be sure it was intended to be used in strict mode or not.

**Comment by Jummit** [comment]
> I think doing it on a per file level makes more sense

The problem OP is having is that he isn't consistent across files. This is will not help solve that.

**Comment by Calinou** [comment]
Ignoring warnings in third-party add-ons should be covered by https://github.com/godotengine/godot/pull/33437, which is now merged and available in 3.2 beta2 and 3.1.2 alike.

**Comment by Shadowblitz16** [comment]
What about statically typing nodes with cylindrical dependencies or errors?
Its currently not possible to do so.

**Comment by Shadowblitz16** [comment]
Looks like gdscript is getting generics and its going to be <> styled which I am fine with since its from C# and c++

**Comment by mirrorlink** [comment]
This is the only reason I still havent migrated to godot. To be fair, a full linter would even be better. So you can decide in which cases you want typing or not, just like typescript.

Maybe godot should support typescript itself? Wouldnt it be much easier? Im pretty sure microsoft would help as they did with C#

**Comment by Calinou** [comment]
> Maybe godot should support typescript itself? Wouldnt it be much easier? Im pretty sure microsoft would help as they did with C#

There are no plans to support more languages officially in Godot, but nothing prevents third parties from creating integrations thanks to GDNative.

**Comment by akien-mga** [comment]
This is probably superseded by https://github.com/godotengine/godot-proposals/issues/3284, but I guess we can keep it open until that one has been implemented and see if we want to take it further (e.g. by enforcing it on a per-file basis instead of a project-wide warning, so that e.g. some plugins could enforce it while the main project would not).

**Comment by TannerBluhm** [comment]
I'm new here, just curious what it would take to move this PR forward?

**Comment by Zireael07** [comment]
@TannerBluhm This is an issue, not a PR.

**Comment by Calinou** [comment]
https://github.com/godotengine/godot/pull/59428 needs to be rebased and reviewed again.

**Comment by TheColorRed** [comment]
I'm still new to Godot (so maybe it's already supported), but when adding strict types wouldn't it be helpful to support union types (#737) otherwise you will have to put `Variant` everywhere, and that isn't very helpful.

```gd
var val: int | float = 123
```

**Comment by Calinou** [comment]
> I'm still new to Godot (so maybe it's already supported), but when adding strict types wouldn't it be helpful to support union types (#737) otherwise you will have to put `Variant` everywhere you want multiple types, and that isn't very helpful.
> 
> ```gap
> var val: int | float = 123
> ```

Godot already allows implicit casts in many situations, such as `float` -> `int`. As a result, union types aren't essential for getting this proposal working.

**Comment by akien-mga** [comment]
Implemented as an optional warning (which can be enforced as error) in https://github.com/godotengine/godot/pull/81355.
